datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -777,29 +848,50 @@ impl Stream for SMJStream {
                     self.state = SMJState::JoinOutput;
                 }
                 SMJState::JoinOutput => {
+                    //dbg!(self.buffered_data.scanning_finished());
+
+                    //dbg!("join_partial");
                     self.join_partial()?;
 
+                    //dbg!(self.buffered_data.scanning_finished());
+
                     if self.output_size < self.batch_size {
                         if self.buffered_data.scanning_finished() {
+                            //dbg!("scanning finished");
                             self.buffered_data.scanning_reset();
                             self.state = SMJState::Init;
                         }
                     } else {
                         self.freeze_all()?;
-                        if !self.output_record_batches.is_empty() {
+                        if !self.output_record_batches.batches.is_empty() {
                             let record_batch = self.output_record_batch_and_reset()?;
+                            let record_batch = if !(self.join_type == JoinType::Left
+                                && self.filter.is_some())
+                            {
+                                record_batch
+                            } else {
+                                RecordBatch::new_empty(Arc::clone(&self.schema))
```

Comment:
this block is magic, but without it the join stucks, I believe it is something with output sizes. We can improve it later

Reply:
Maybe you could change it to `continue` to avoid outputting an empty batch 🤔